### ISSUE TITLE
Android NDK Clang produces 23% slower binaries than GCC

### ISSUE THREAD

**Issue by bruenor41** [android-ndk/ndk#495]
--------------------------------------------------------------------------------
Switching from android gcc to clang produces slower binaries.

#### Description

I am working on application what emulates old operating system from 90'. I build c++ binaries with android ndk r10e with gcc 4.8. But, because google drops support for gcc, I want switch to clang. After updating ndk to r15 and successful build, I ran benchmarks(in r15 is default compiler clang, instead gcc like it was in ndk r10e). Result is that emulated system runs 23% slower, even without benchmark it is very noticeable when playing more cpu demanding games.

I did none changes to Android.mk and I use -O3 for optimizations. I only updated ndk. Simply said, I only switch between toolchains with : 
NDK_TOOLCHAIN_VERSION=4.9 or without it

In ndk r15 is still possible switch to gcc 4.9. I did it and I got back lost performance. And even more, seems gcc 4.9 makes slightly better optimizations then gcc 4.8. 

From what I read on web I expected that clang will produce faster binaries or the same. And I didn't find special magic clang flags what I must enable for Android.mk.

For a test I returned back to ndk r10e and switched to clang 3.8. Compiled application is again around 20%+ slower. Seems it is not related to ndk r15, but is there long time.

#### Environment Details

In my application.mk I use this setup 

APP_ABI := armeabi armeabi-v7a x86
APP_OPTIM := release
APP_STL := stlport_static
APP_PLATFORM := android-8
LOCAL_LDLIBS += -lz 

APP_PLATFORM is set to 8 but I see in log that minimum was set to 14 automatically

Android.mk
APP_CPPFLAGS = -O3
LOCAL_ARM_MODE := arm
LOCAL_CFLAGS += -DHAVE_NEON=1
LOCAL_ARM_NEON := true

I tested on windows 10 and ubuntu 16.04, results are same
Tested on device with android 6.0 and 7.0

**Comment by stephenhines** [comment]
This isn't really that actionable without a repro case. 20%+ slower seems outside of the bounds of anything we have ever seen (in 7+ years of working on this), but it's hard to do more without further information.

**Comment by bruenor41** [comment]
I understand.  I can create application what you can test, it takes me 1-2 days. But I don't like make sources public. I can provide them via email to someone from team.

**Comment by kneth** [comment]
Isn't `-O3` a bit adventurous (https://github.com/android-ndk/ndk/issues/110#issuecomment-226261466)?

**Comment by bruenor41** [comment]
Unfortunately I tried all these flags, but none brings back performance
https://stackoverflow.com/questions/15548023/clang-optimization-levels

**Comment by enh** [comment]
have you tried adding -Bsymbolic? that's one of the main differences between the default clang flags and the default [Android] GCC flags.

**Comment by bruenor41** [comment]
nope, today is feast in our country, but will try tomorrow immedately at morning. Thank you very much for idea

**Comment by bruenor41** [comment]
I created benchmark [application](https://www.dropbox.com/s/lzt8esdo8uwqkwg/BenchmarkTest.zip?dl=0). Please check readme.txt in root

**Comment by stephenhines** [comment]
A couple of quick suggestions. I am not sure whether APP_CPPFLAGS (where you set -O3) gets cleared by $(CLEAR_VARS). I will also note that there are several .c source files in this example, which will not get the -O3, since that would require setting APP_CFLAGS instead (which applies to both .c and .cpp files). Instead. these files will be compiled at the Clang default (-O0), which is bound to have performance problems. Since I don't really work on the NDK build side of things, this is my first experience with any of these special make variables, so I am really guessing that not all the flags you expect are being passed to your actual built libraries. LOCAL_CFLAGS is much more familiar to me, and those do get cleared by $(CLEAR_VARS).

**Comment by bruenor41** [comment]
Stephen, thank you very much  for suggestions! I replaced APP_CPPFLAGS with LOCAL_CFLAGS. It was definitely good idea, binaries made by both compilers profits from his change and performs slightly better. Each performance gain is for me important. Unfortunately gcc still performs much better :(

**Comment by stephenhines** [comment]
You need to update every place after $(CLEAR_VARS) to add a LOCAL_CFLAGS with your optimization level to have it affect that compilation. Otherwise, I am pretty convinced at this point you are getting -O0 compilation for many of the sources files with Clang.

**Comment by bruenor41** [comment]
[Here](https://www.dropbox.com/s/8588fwr00d1njm7/mk_files.zip?dl=0) is how I updated them

**Comment by DanAlbert** [comment]
Could you just upload that as a github project or at least a gist so we don't have to keep redownloading it?

**Comment by DanAlbert** [comment]
`APP_CFLAGS` (or any other `APP_*` variables) does nothing in Android.mk. Application-wide flags need to be set in Application.mk.

**Comment by bruenor41** [comment]
here[https://github.com/bruenor41/Benchmark](url)

**Comment by DanAlbert** [comment]
Which ABI is the one you're having issues with?

**Comment by bruenor41** [comment]
armeabi-v7a. I don't own x86 based device, I can test it in emulator, but I didn't. And old armeabi I have only due possible fallback, it is not very used.

**Comment by bruenor41** [comment]
Please has someone idea what might be a problem? I like to switch to clang as soon as possible

**Comment by DanAlbert** [comment]
I'll do what I can to minimize a test case for the compiler folks.

**Comment by bruenor41** [comment]
Thank you very much, do not hesitate contact me if you need something

**Comment by DanAlbert** [comment]
How do I build your project?

**Comment by DanAlbert** [comment]
Never mind, I reopened it and Studio stopped complaining.

**Comment by DanAlbert** [comment]
Confirmed a 12% perf regression for clang vs gcc on Pixel 2.

Can you use [simpleperf](https://developer.android.com/ndk/guides/simpleperf.html) to try to reduce this? I didn't realize the "benchmark" was a full game... If you can get us something more concrete to work with (and something that builds out of the box) then we can take a look, but this beyond the scope of what we'll be able to look at any time soon I think.

**Comment by bruenor41** [comment]
Hi Dan,

I never heard about simpleperf, but will try it next week and will post
result.

I already simplified complex emulator, now it starts automatically this
benchmark - game which counts how fast is host pc, in our case it means how
fast runs emulator

I am standard programmer (c, c++, java,..) I dont understan low level code.
I didnt make changes in code, but program compiled with clang runs slower.
I see this issue in competitive programs too - in programs which already
use clang. I hoped that ndk developers can compare *.o files, or maybe they
can run some "magic" profiler, sorry for my naivety..:)

**Comment by bruenor41** [comment]
I was thinking about it. Maybe I can give some hints and minimize the
problematic area. I believe it could be something with dynamic recompiler
to arm.

90% of all emulated power is about cpu emulation. Cpu emulation is really
sensitive, based on device cpu (device on which I run emulator).

Sources for cpu are located in jni/dosbox/src/cpu.
For arm are important files : cpu.cpp, core_dynrec.cpp
Then, is enabled dynamic recompiler to arm controlled with
DC_TARGETCPU=ARMV7LE.
If you search this flag then you can find appropriate plugin
risc_armv4le-o3.h

**Comment by ehem** [comment]
Might this this be related to #21?

**Comment by stephenhines** [comment]
This is probably not related to #21. As I mentioned above, this project is a combination of many libraries, so it is very likely that there are other places where optimization flags are not being set or configured properly. I found one by quick inspection, and that already improved things. Diagnosing the rest of this will require someone to spend a lot more time with simpleperf and the entire build. I will note that this is not a high priority thing for my team to do, as our platform benchmarks don't show similar regressions (in performance or code size).

**Comment by bruenor41** [comment]
You are wrong Stephen. This project is very standalone. Is used "stlport_static" in application.mk, but else all source codes used are written in android.mk file and are affected with compiler flags. 

Your information **"it is very likely that there are other places where optimization flags are not being set or configured properly. I found one by quick inspection, and that already improved things"** is old, because I updated android.mk like you mentioned and all *.c and *.cpp files are affected with LOCAL_CFLAGS.

In addition, previous or currently modified android.mk file creates faster binaries with gcc then with clang, I believe LOCAL_CFLAGS has zero effect. It must be something else. Probably changes in stlport_static (or is untouched since ndk r10e?)

In addition2, I can name at least 1 competitive application in production on google play which suffers the same issue from other author and with other build system.

**Comment by DanAlbert** [comment]
Yeah, there wasn't anything obviously wrong with the build when I looked at it.

**Comment by bruenor41** [comment]
I updated git repo, I removed some stuff not necessary for test like sdl thread or ogg support

**Comment by bruenor41** [comment]
Well, I found another strange thing. I was trying to find what can cause this issue. I though that problem may be in arm plugin, so I tried compile for x86 and I wanted compare binaries made by clang and gcc. I use android emulator on API 26 and 25 (two different emulated devices) with enabled haxm. App made by gcc runs fine, app compiled with clang **hangs** soon after start.

**Comment by bruenor41** [comment]
Hi, I am trying run simpleperf, but on soon after start "python app_profiler.py" I get error message :

DEBUG:root:run adb cmd: ['adb', 'shell', 'run-as', 'benchmarkdos.doom', 'cat perf.data | tee /data/local/tmp/perf.data >/dev/null']
DEBUG:root:/system/bin/sh:  tee /data/local/tmp/perf.data : not found
run-as: exec failed for cat perf.data : Permission denied

Does simpleperf requires rooted device? Or is that error related to something else?

**Comment by DanAlbert** [comment]
@yabinc or @enh should be able to answer that. I've never used it.

**Comment by yabinc** [comment]
I profiled the Benchmark app today. The result is in https://github.com/yabinc/profiling_data/blob/master/Benchmark/summary.
The profiling result is shown in report.html.
It seems at least partly because gcc is more aggressive at inlining code.

@bruenor41, For the problem of running app_profiler.py:
DEBUG:root:run adb cmd: ['adb', 'shell', 'run-as', 'benchmarkdos.doom', 'cat perf.data | tee /data/local/tmp/perf.data >/dev/null']

I don't know why it can happen, but I suggest you try the simpleperf in the latest ndk release r16 Beta2,
at https://developer.android.com/ndk/downloads/index.html#beta-downloads.
If you experience any problem, please feel free to raise an issue in ndk issues.

**Comment by bruenor41** [comment]
Thank you @yabinc for help, I really appreciate it. Really big thanks.

**Comment by ehem** [comment]
After fighting an issue I found this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66917

Apparently the Linux kernel has been patching this at runtime for some time now.  Depending upon how much memcpy() is used by a given piece of code, those speed estimates for GCC might well be highly conservative since this could be a major speed issue for some pieces of code.

**Comment by kneth** [comment]
@ehem Interesting. I had a similar experience, and rolled my own `memcpy` as a workaround. See https://www.slideshare.net/geisshirt/tales-from-the-dark-side-developing-sdks-at-scale - slides 14-22

**Comment by pradeep250677** [comment]
@yabinc , as per your benchmark , inline may be  the issue. clang have some optimization issue with "r,m" constraint syntax  to force use register which GCC managed very well. I do not know the root cause of clang .
We can try static inline instead of inline function. It may be help to improve performance .

**Comment by bruenor41** [comment]
Currently I am still at ndk r10e. Customers will crucify me if performance
will drop so much :) Fortunately, I can still be at that old ndk, because
currently I am not sure how to fix this problem. Hope it will be fixed in
one of next clang versions. Or if someone with better knowledge can help ..

**Comment by pcorvodts** [comment]
Just to chime in and plus one on this we are seeing a very similar performance regression moving up ndks. We upgraded to ndk r16b (clang) from ndk r12 (gcc) and see a regression of a very similar amount on our  arm32 libraries. It doesn't happen on our ARMv8 libraries (which actually benefit from a considerable performance improvement on the new compiler) but there is a definite ~20% regression on arm32. We're currently investigating but have no clues yet :(

**Comment by dimitryn** [comment]
We face the same behaviour when moved all our c++ libraries to NDK 17.
OpenCV, Tensorflow, Tensorflow Lite all compiled with NDK 17 and all of then now running much slower.
can anyone update if they succeed resolve this issue? @bruenor41 @DanAlbert @stephenhines @kneth @enh

**Comment by bruenor41** [comment]
I did not solve this problem, I am stuck. However, your trouble seems to be different. I report performance degradation in all builds made by clang. This was still not fixed. Switching to gcc fixed this issue in all ndk's before ndk17. However, in ndk 17 something strange has happen, because both clang and gcc builds suffers from big performance degradation.

Meanwhile was my app ported and now supports arm8, I will do new reports soon, because it's one year since this report and it still was not fixed.

**Comment by DanAlbert** [comment]
> Switching to gcc fixed this issue in all ndk's before ndk17. However, in ndk 17 something strange has happen, because both clang and gcc builds suffers from big performance degradation.

Just to clarify: did you try r16? You say all, but just want to make sure because I would actually expect that change to happen in r16 if at all. r16 brought GCC's default behavior in line with its behavior on every other platform (and also Clang) by no longer defaulting to `-Bsymbolic`. This affects symbol preemption which presumably could be affecting the compiler's inlining decisions. The way to test this would be to add `-Bsymbolic` to your ldflags and see if you see the old behavior return.

**Comment by DanAlbert** [comment]
If that's not it, then I'm not sure what would have caused r17's GCC behavior to change, as we did not update GCC in that release (we technically did because we updated libgcc to fix some bugs, but no changes were made to GCC proper).

**Comment by bruenor41** [comment]
Thank you Dan, I did these tests some time ago, I was sure that it happen by ndk 17, but I'll check it again tomorrow and I will write result.

**Comment by DanAlbert** [comment]
(It's also possible that I'm forgetting some r17 default flag change or something similar. It was a fairly long release cycle and my memory is crap. I can dig deeper if you find that it was r17 and not r16 that caused that change.)

**Comment by pcorvo83** [comment]
Hello, I thought I had actually contributed to this thread before but can’t find my comment. I think I might have thought about replying but wanted to re-check our pipeline first (which is fairly complex) before adding a comment.

Essentially we found the same thing when upgrading from ndk r12 (gcc default) to ndk r16 (clang default). We build an SDK and actually saw a big performance improvement in our ARMv8a 64bit libraries but a 20-30% performance regression on our ARMv7a 32bit libraries. 

In the end our current workaround was to change our build system so that we build arm64 libraries with clang on ndk16 and arm32 libraries with gcc on ndk16 to give us the improvements in arm64 libs and no regression on arm32 libs.

We would like to get to the bottom of this but it went in the backburner for a while while we have a solution. I’m yet to check disassemblies for obvious differences but would be curious to hear of other peoples experiences.

**Comment by bruenor41** [comment]
Hi Dan, I did benchmark for 32 bit libraries (armeabi-v7a), here is result. Lower number means better performance. 

**NDK 15**
gcc 1900
clang 2249

**NDK 16b**
gcc 2020
clang 2255

**NDK 17b**
gcc 2050
clang 2240

While clang makes binaries constantly slow, gcc binaries are much slower since ndk 16

**Comment by DanAlbert** [comment]
Did you try adding `-Bsymbolic` to your ldflags? That should both improve clang performance return GCC performance to its pre-r16 levels then.

**Comment by dimitryn** [comment]
https://software.intel.com/en-us/articles/performance-tools-for-software-developers-bsymbolic-can-cause-dangerous-side-effects @DanAlbert  have you read this

**Comment by enh** [comment]
"don't do any of those silly things and you'll be fine".

most of them aren't possible on Android anyway, since you're not duplicating types/data between your .so and ART.

**Comment by DanAlbert** [comment]
@dimitryn: Yes, but that's not really relevant to this topic. `-Bsymbolic` was implicit in GCC until r16. It was a bad default since it was something unique to Android GCC (no other platform nor compiler does this).

The reason that point is important for this thread is that pre-r16 GCC and Clang comparisons are invalid. You need to compare r16+ for an even comparison.

**Comment by bruenor41** [comment]
@DanAlbert I added 

LOCAL_LDFLAGS := -Bsymbolic

to armv7.mk file and to Android.mk for magiclib. For all libs. Hope the syntax is correct. But no luck. I tested ndk 17 and here is result :

**NDK 17b**
clang 2195
gcc 2020

gcc is still on same value, clang seems to be a bit faster. Or if the syntax is not correct, could you fix both mk files and post them? 

Thank you

**Comment by Murphy0571** [comment]
Guys,

Is there any conclusion in this topic? Performance decreasing is also seen in my side, but not a clue yet :'(

Br, Murphy

**Comment by arturbac** [comment]
I use clang on linux for few years , and android about 1 year, i didint notice performance drops with clang >= 4.x, 
have You tryied:
-Ofast3 with clang - enables fast-math optimisations, but may be dangerous to old invalid code
-fno-stack-protector - disable stack protector strong - SP causes that args are passes on stack  instead of registers
-fintegrated-as - use llvm as instead on binutils
-fuse-ld=lld -flto=thin - enable thin lto and llvm ld  (thin-lto works on linux with ndk18 beta1)

**Comment by bruenor41** [comment]
Hi, 19.7 I tried ndk 17b, in ndk changelog is written :
Updated Clang to build 4691093, based on r316199.

Maybe is not problem in clang self, maybe clang uses some resources which are not used by gcc. This is question on ndk team. I don't know. However, demo for testing was included one year ago, but seems there is no interest.

**Comment by arturbac** [comment]
Up in topic there is some asm code comaprision, this so strong difference must be achived only by way different compilations flags, like stack protector strong, which was enabled in latest ndks, later it as only -fsatck-protector. I worked for a while with disasemble of code, during bug hunting. SP causes a lot of deoptymalizations of arguments, like not passing in registers, doing a lot of indexed memory accesses. If you app is stable , not causing stack buffer overflows it is worth thinking if you really need that feature (depending on app type,ex game or financial apps). SP also causes harder cracking. if performance isnt the case it is nice feature, but for cpu intensve app it is a killer, whiel for ex financial app it is an another security protection.

**Comment by bruenor41** [comment]
Thank you very much for explanation. Unfortunately I am not able debug assembler, because I do not understand it :( My application is emulator and needs a lot of cpu power. I simply switched from gcc to clang like was recommended by google and immediately I noticed more then 20% performance degradation in emulation. 

@yabinc made deep comparison of gcc and clang made binaries here :
https://github.com/yabinc/profiling_data/blob/master/Benchmark/summary

This report helped me to optimize some routines. This affected binaries made by both compilers are now now faster. However main problem I did not solve...

I think it's more then I can hiope, but if there is little hope to look at my benchmark demo then I will be very happy

**Comment by arturbac** [comment]
1) compiled binary will not help, there arent any information about compialtion flags, ex from cmake CMAKE_C_FLAGS CMAKE_CXX_FLAGS
2) I can excamine compilation result if You insert into gh repo - one function that can be compiled independently, some low level function that is slow.
3) There could be another thing with old code diged out from some old emulators - but it usualy causes crashes with modern compilers. Memory aliasing. modern compilers assume no aliasing of types to single memory location, this gives a lot of speedup but in old times it was common to alias memory and access it but reinterpret cast/C hard cast from one to other type.Clang by default assumes no alising, whiel old gcc 4.9 assumed that aliasing may happen without switch -strict-aliasing.

**Comment by Murphy0571** [comment]
@arturbac Your comments are a real asset! I will try your recommended compile flags and see if it can make the magic here!

**Comment by arturbac** [comment]
2 things that should be changed in this project
- remove armeabi, add arm64-v8a
- switch from mk build system to cmake

**Comment by bruenor41** [comment]
Yes, armeabi is obsolete. arm64 dynamic recompiler is finished and will be added. I did not test it with clang and gcc yet. Maybe it will be better. However I report 32bit.

Could there be differences in resulted binaries using cmake?

**Comment by arturbac** [comment]
1) armv64-v8a on android must be build at least with android 5.x api 21, there werent older 64bit androids.
armv7a with ndk17 assumes ndk 14 (adnroid 4.x)
2) as i mentioned in 64bit build cpu has available 2 times more number of registers, this is significant in optimalizations and generated code
3) cmake - s i know mk i deprecated build system, android-toolchain.cmake is upgraded continiuusly, that one thing decides what default compilations flags are used, what libs are linked. Dont know if there is difference in mk vs android-toolchain.cmake, if both are still upgraded. Staying on mk build system dosnt make sense anyway.
Hovever both systems manualy configured can generate same compialtion options.
And that options passed to compiler affects generated code most.

**Comment by arturbac** [comment]
BTW in 64bit mode you will get also advanced simd instructions but more important by default integer division is done with hardware while in armv7a idiv is done by software impl by default, as not all armv7a cpus has got not mandatory for armv7 cpu extnsions idiva idivt

**Comment by bruenor41** [comment]
Thank you. That's reason why arm64 dynamic recompiler was made. However, I must still compile for 32 bit devices...

**Comment by arturbac** [comment]
1) I looked at the perf result from clang

The host spot is in very ineefficient method

86 | 27.000 ms | 27.000 ms | bool PageHandler::writeb_checked(PhysPt addr,Bitu val) {
-- | -- | -- | --
87 | 2298.000 ms | 173.000 ms | writeb(addr,val);	return false;
88 |   |   | }

every single write is done by
- calling function
- which dereferences address
- stores the result
-returns
this one should be moved at least to header or build this with LTO - thin
bool PageHandler::writeb_checked(PhysPt addr,Bitu val) {
	writeb(addr,val);	return false;
}

2)  gcc profiling has totaly diffrent hostspot functions
- It looks like that some #defines in code are enabled/disabled when compiler is gcc, it looks like some code during clang compialtion is using checked version of implementation which propably should be enabled only for debuging purproses. the checked tag in function name and its inefficient implementation in .cpp from clang sugests that too.

when i diged out what readb does
i saw in asm blx instruction , conditional branch during write
in source there is
void PageHandler::writeb(PhysPt addr,Bitu /*val*/) {
	E_Exit("No byte handler for write to %x",addr);	
}

so it looks like for clang E_exit does somthing cpu hungry (loging or what ever ) which is disabled on gcc release ..

**Comment by enh** [comment]
One note: ndk-build is _not_ deprecated. It's every bit as supported as cmake is.

For the most part, the two should behave the same, and where they don't you should feel free to report bugs. Work in NDK r19 to move logic out of the build systems and into the clang driver should also help here.

**Comment by DanAlbert** [comment]
@bruenor41: sorry, I'd missed this update:

> @DanAlbert I added
> 
> LOCAL_LDFLAGS := -Bsymbolic
> 
> to armv7.mk file and to Android.mk for magiclib. For all libs. Hope the syntax is correct. But no luck. I tested ndk 17 and here is result :
> 
> NDK 17b
> clang 2195
> gcc 2020
> 
> gcc is still on same value, clang seems to be a bit faster. Or if the syntax is not correct, could you fix both mk files and post them?
> 
> Thank you

I wonder if the flags aren't being added in the right order. You could add `-v` to your `LOCAL_LDFLAGS` (`APP_LDFLAGS` would be even better). That would also help you figure out what changed for GCC between the two releases. `-v` will show you exactly how the compiler internals and the linker were invoked. Diff the two and see what changed.

**Comment by bruenor41** [comment]
Hi people. After long time, I was able compile this benchmark for arm64-v8a, so I could compare performance against armeabi-v7a. I used ndk 15 and I compared clang and gcc binaries. And I was surprised! while arm7 binaries made by clang reports so big performance degradation, arm8 binaries gives almost same result. Performance is slightly better with gcc binaries, around 3%. 

arm7 and arm8 comparison took so long time, because benchmark uses dynamic recompiler which was not made for arm8 before, now is.

**Comment by DanAlbert** [comment]
Reading back through this thread, it sounds like the differences discussed here are a result of 1. differences in inlining behavior and 2. intentional differences in behavior between clang and gcc within the project. The first is a trade-off where Clang has made a different decision than GCC, and the latter needs to be fixed in the project itself, not Clang.

> arm7 and arm8 comparison took so long time, because benchmark uses dynamic recompiler which was not made for arm8 before, now is.

If I'm understanding this correctly, it sounds very plausible that the project's recompiler has just been optimized thoroughly for 32-bit but not for 64-bit.

Given that, I don't think there's any action to be taken here.

**Comment by bruenor41** [comment]
No Dan. Previously I compared 32 bit binaries made by clang against 32 bit binaries made by gcc. Gcc performs much better in execution speed. I removed all gcc flags from, but it did not help. Gcc still optimizes it better. 

Now I compared 64 bit binaries made by gcc against 64 bit binaries made by clang. Performance is roughly the same, why. In this case I reverted back sources with gcc flags, so they were included during test.

**Comment by DanAlbert** [comment]
I've spent the last few hours just trying to get your project to build. I finally succeeded, but it crashes when I start it.

I stand by what I said above. The fact that you do see the regression with GCC in newer NDKs makes me very confident that this is not a Clang issue. If you really want to dig in to what changed between r16's and r17's GCC, add

```makefile
APP_CFLAGS := -v
APP_LDFLAGS := -v
```

to your Application.mk files and rebuild. gcc will show the flags that it is using to invoke the compiler, the assembler, the linker, etc. Look for changes in the flags used between your old GCC build and a new one. Odds are that will have your answer.

This is what I was trying to do while trying to build your project, but apparently one successful build is all I get. After cleaning the project, the *Java* code doesn't even compile.

**Comment by bruenor41** [comment]
Thank you very much for your digging, I really appreciate it. My problem is not that gcc has been changed between ndk 16 and ndk 17 and that it's getting yet worse. My problem is that I need to switch to clang and project compiled by clang makes much slower binaries since beginning - in my case beginning means ndk r10e, I did not test lower versions. 

Please, I am not sure why you have problems compile it, but I can look it this in in next hour. I will be very happy for any clue.

**Comment by DanAlbert** [comment]
Updating studio and Gradle was the first thing I had to do, but I also had to make changes to get things working with `externalNativeBuild`. Without that it's very difficult to get useful information from Studio. It's something you'll want to do for your project regardless.

**Comment by bruenor41** [comment]
Ah I know now, sorry. I was releasing new version and I did not realize that my app uses still ndk r10e. The same for github version of benchmark. I updated to newer ndk only on local repository, I will fix that

**Comment by bruenor41** [comment]
Hi Dan, I updated project, you shouldn't have problems now. Please, I will be thankful for any kind of help. I spent a lot of time on debugging and switching between clang and gcc flags or removing them. I do not know on what I can focus now, except debugging assembler what is behind my knowledge...

**Comment by DanAlbert** [comment]
Thanks. It won't be taking priority over other work but I'll keep in my list of Friday projects.

**Comment by bruenor41** [comment]
Yes, I understand, no hurry. Many thanks

**Comment by ZaccurLi** [comment]
Is there any conclusion in this topic?  .
We face the same issue when switched all our c++ libraries from gcc to clang.
Performance decreasing also occured.
OpenCV, and our source code all compiled with NDK 17 were running much slower.
can anyone update if they succeed resolve this issue? Thanks.

**Comment by noyanc** [comment]
Hi, we also face the same issue after switching from gcc to clang.

Old framerate of the game was 55 FPS. After passing to clang, the framerate dropped to 23 FPS on the same device. Over 55% performance loss! (We build against ndk-r19c using -Ofast3 and all other additional optimization flags.)

Since clang and 64 bits compiling is mandatory, this issue has become urgent for my team!

Would be great if any progress.

**Comment by bruenor41** [comment]
I like to help, but i did not fix it. I did what i could to solve this
issue, but i did not succeed and there was enough interest on ndk
developers side

**Comment by joeshow79** [comment]
Hey, we found the downgrade  of the performance seems introduced by openmp.
Take the following test sample as example
main.cc

`int main() {                                                                                                                                                                                           
    int sum;                                                                                                                                                                                                      int N = 100000000;                                                                                 
#pragma omp parallel for reduction(+: sum) schedule(dynamic,10)                                        
    for(int i=0; i<N;i++){
        sum += i;
    }       
} `

Makefile:
`all:test_gcc test_clang test_gcc_omp test_clang_omp

test_gcc_omp:main.cc
    g++ -fopenmp -o $@ $^
                                                                                                                                                                                                              
test_clang_omp:main.cc
    clang++ -fopenmp -o $@ $^

test_gcc:main.cc
    g++ -o $@ $^

test_clang:main.cc
    clang++ -o $@ $^`

test script:
`echo "test gcc version:"
time ./test_gcc

echo "---------------------------------"

echo "test clang version:"                                                                                                                                                                                    
time ./test_clang

echo "---------------------------------"

echo "test gcc omp version:"
time ./test_gcc_omp

echo "---------------------------------"

echo "test clang omp version:"
time ./test_clang_omp`

The result seems as below for several running of the test  (test in linux docker without NDK)
`bash test.sh 
test gcc version:

real    0m0.236s
user    0m0.236s
sys     0m0.000s
---------------------------------
test clang version:

real    0m0.233s
user    0m0.232s
sys     0m0.001s
---------------------------------
test gcc omp version:

real    0m0.243s
user    0m0.469s
sys     0m0.015s
---------------------------------
test clang omp version:

real    0m0.331s
user    0m0.651s
sys     0m0.008s`

It shows that the clang is even a little bit faster than the gcc counterpart if omp is disabled. However the performance degrade drastically when openmp is enabled . Similar result observed for the test on Android(with NDK) from the flame chart produced by other test. Can not paste the screenshot of the flam chart here for more detail.

So may anybody can dig further to see what the clang openmp done behind the scene?

**Comment by tpetri** [comment]
Hi,

I was following this thread now for many years in hope for a solution, we have the exact same experience as many here. Code compiled with Clang with omp for Android does result in really bad performance. On other platforms we cannot see these issues.

Does anyone had success in finding a solution/reason/workaround? Would it be possible for the NDK devs to maybe try to have a look again?

**Comment by MoNTE48** [comment]
Android developers just don't care at all. Forget it is closed because it will never be fixed. Buy new devices. Old don't bring money

**Comment by arturbac** [comment]
Check first the source code of bottleneck that has significant difference, then check in that code if there are specific macro definitions for gcc that causes to generate extraordinary optimized code for example for SIMD.
I can say that code generated with clang is in 99% at least as god as generated with gcc, there are many situations when clang generates much better code than gcc! for aarch64.
Use compiler explorer for parts of code and compare generated code by clang and gcc.
It will be very hard for You to make a prof that clang generates at least a little slower code.

**Comment by arturbac** [comment]
> int main() { int sum; int N = 100000000; #pragma omp parallel for reduction(+: sum) schedule(dynamic,10) for(int i=0; i<N;i++){ sum += i; } }

This test dsont make any sense because:
1) You are compiling _without_ any specified optimizations
2) With optimized version ex -Ofast3 You will get with clang optimise out entire code as result of ccomputation is unused
main:                                   // @main
        mov     w0, wzr
        ret

**Comment by joeshow79** [comment]
> Switching from android gcc to clang produces slower binaries.
> 
> #### Description
> I am working on application what emulates old operating system from 90'. I build c++ binaries with android ndk r10e with gcc 4.8. But, because google drops support for gcc, I want switch to clang. After updating ndk to r15 and successful build, I ran benchmarks(in r15 is default compiler clang, instead gcc like it was in ndk r10e). Result is that emulated system runs 23% slower, even without benchmark it is very noticeable when playing more cpu demanding games.
> 
> I did none changes to Android.mk and I use -O3 for optimizations. I only updated ndk. Simply said, I only switch between toolchains with :
> NDK_TOOLCHAIN_VERSION=4.9 or without it
> 
> In ndk r15 is still possible switch to gcc 4.9. I did it and I got back lost performance. And even more, seems gcc 4.9 makes slightly better optimizations then gcc 4.8.
> 
> From what I read on web I expected that clang will produce faster binaries or the same. And I didn't find special magic clang flags what I must enable for Android.mk.
> 
> For a test I returned back to ndk r10e and switched to clang 3.8. Compiled application is again around 20%+ slower. Seems it is not related to ndk r15, but is there long time.
> 
> #### Environment Details
> In my application.mk I use this setup
> 
> APP_ABI := armeabi armeabi-v7a x86
> APP_OPTIM := release
> APP_STL := stlport_static
> APP_PLATFORM := android-8
> LOCAL_LDLIBS += -lz
> 
> APP_PLATFORM is set to 8 but I see in log that minimum was set to 14 automatically
> 
> Android.mk
> APP_CPPFLAGS = -O3
> LOCAL_ARM_MODE := arm
> LOCAL_CFLAGS += -DHAVE_NEON=1
> LOCAL_ARM_NEON := true
> 
> I tested on windows 10 and ubuntu 16.04, results are same
> Tested on device with android 6.0 and 7.0

Is openmp  used in the code, we figured out there are some 

> Hey, we found the downgrade of the performance seems introduced by openmp.
> Take the following test sample as example
> main.cc
> 
> `int main() { int sum; int N = 100000000; #pragma omp parallel for reduction(+: sum) schedule(dynamic,10) for(int i=0; i<N;i++){ sum += i; } } `
> 
> Makefile:
> `all:test_gcc test_clang test_gcc_omp test_clang_omp
> 
> test_gcc_omp:main.cc
> g++ -fopenmp -o $@ $^
> 
> test_clang_omp:main.cc
> clang++ -fopenmp -o $@ $^
> 
> test_gcc:main.cc
> g++ -o $@ $^
> 
> test_clang:main.cc
> clang++ -o $@ $^`
> 
> test script:
> `echo "test gcc version:"
> time ./test_gcc
> 
> echo "---------------------------------"
> 
> echo "test clang version:"
> time ./test_clang
> 
> echo "---------------------------------"
> 
> echo "test gcc omp version:"
> time ./test_gcc_omp
> 
> echo "---------------------------------"
> 
> echo "test clang omp version:"
> time ./test_clang_omp`
> 
> The result seems as below for several running of the test (test in linux docker without NDK)
> `bash test.sh
> test gcc version:
> 
> ## real 0m0.236s
> user 0m0.236s
> sys 0m0.000s
> test clang version:
> 
> ## real 0m0.233s
> user 0m0.232s
> sys 0m0.001s
> test gcc omp version:
> 
> ## real 0m0.243s
> user 0m0.469s
> sys 0m0.015s
> test clang omp version:
> 
> real 0m0.331s
> user 0m0.651s
> sys 0m0.008s`
> 
> It shows that the clang is even a little bit faster than the gcc counterpart if omp is disabled. However the performance degrade drastically when openmp is enabled . Similar result observed for the test on Android(with NDK) from the flame chart produced by other test. Can not paste the screenshot of the flam chart here for more detail.
> 
> So may anybody can dig further to see what the clang openmp done behind the scene?

We finally found there has something to do with the OpenMP configuration, we made the following changes in code and seems the clang and gcc compiled binaries are finally at the close performance level.

KMP_BLOCKTIME=0
OMP_WAIT_POLICY=PASSIVE

**Comment by tpetri** [comment]
Thank you @joeshow79 for the hint!

This actually improves the situation a lot! No more strange thread blocking. Just as a remark, on certain phones (e.g. Honor View 20) there seems to be an OpenMP version that listens to the GNU environment variables, I had to set GOMP_SPINCOUNT to 0 to get the same result.

What we did: in Application onCreate() before loading any native libs:
```
Os.setenv("OMP_WAIT_POLICY", "passive", true);
Os.setenv("KMP_BLOCKTIME", "0", true);
Os.setenv("GOMP_SPINCOUNT", "0", true);
```

What I noticed as well, now when giving the Clang-OpenMP issue a new try: on certain phones this was fixed after some Android update. The manufacturers were informed about this or noticed it themselves or whatever. Anyways, this is the solution for us, thanks again.
